### PR TITLE
Introduce Azure Maps cmdlets

### PR DESCRIPTION
**Pull request by jp94**
## Description
Introduce Azure maps cmdlets. 

Most are renaming LocationBasedServices cmdlets to Maps. 
However, I had to attach a [Microsoft.Azure.Management.ResourceManager.ResourceManagementClient](https://github.com/Azure/azure-powershell/pull/6259/commits/29d1123aaf7143a612c1717775bcd7ceed8d6274#diff-715b14ab5f3c3a05ac23bb30390b254bR32) to pass test cases.

Cmdlet review:
https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/21

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

File: src/ResourceManager/Maps/Maps.Test/ScenarioTests/MapsAccountTests.ps1 [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        # Call create again, expect to get the same account
+        $createdAccountAgain = New-AzureRmMapsAccount -ResourceGroupName $rgname -Name $accountname -SkuName $skuname -Force;
+        Assert-NotNull $createdAccountAgain
+        Assert-AreEqual $createdAccount.Id $createdAccountAgain.Id;
+        Assert-AreEqual $createdAccount.ResourceGroupName $createdAccountAgain.ResourceGroupName;
+        Assert-AreEqual $createdAccount.Name $createdAccountAgain.Name;
+        Assert-AreEqual $createdAccount.Location $createdAccountAgain.Location;
+        Assert-AreEqual $createdAccount.Sku.Name $createdAccountAgain.Sku.Name;
+        
+        Retry-IfException { Remove-AzureRmMapsAccount -ResourceGroupName $rgname -Name $accountname -Confirm:$false; }
```
Comment (maddieclayton) [inline src/ResourceManager/Maps/Maps.Test/ScenarioTests/MapsAccountTests.ps1]: Please add a check here to ensure the account was deleted (make sure get returns nothing)

File: src/ResourceManager/Maps/Maps.Test/TestController.cs [hunk 2]
```diff
@@ -182,9 +187,14 @@ private GalleryClient GetGalleryClient()
             return TestBase.GetServiceClient<GalleryClient>(this.csmTestFactory);
         }
 
-        private Client GetLocationBasedServicesManagementClient(RestTestFramework.MockContext context)
+        private MapsManagementClient GetMapsManagementClient(RestTestFramework.MockContext context)
+        {
+            return context.GetServiceClient<MapsManagementClient>(RestTestFramework.TestEnvironmentFactory.GetTestEnvironment());
+        }
+
+        private LegacyResourceManagementClient GetLegacyResourceManagementClient(RestTestFramework.MockContext context)
```
Comment (maddieclayton) [inline src/ResourceManager/Maps/Maps.Test/TestController.cs]: What changed to mean that you need this? Did you change the tests?

File: src/ResourceManager/Maps/Maps.Test/packages.config [hunk 3]
```diff
@@ -5,18 +5,20 @@
...
   <package id="Microsoft.Azure.Management.Storage" version="3.0.0" targetFramework="net45" />
   <package id="Microsoft.Azure.Test.Framework" version="1.0.6179.26854-prerelease" targetFramework="net45" />
-  <package id="Microsoft.Azure.Test.HttpRecorder" version="1.6.7-preview" targetFramework="net45" />
+  <package id="Microsoft.Azure.Test.HttpRecorder" version="1.8.1" targetFramework="net452" />
   <package id="Microsoft.Bcl" version="1.1.9" targetFramework="net45" />
   <package id="Microsoft.Bcl.Async" version="1.0.168" targetFramework="net45" />
   <package id="Microsoft.Bcl.Build" version="1.0.14" targetFramework="net45" />
   <package id="Microsoft.IdentityModel.Clients.ActiveDirectory" version="2.28.3" targetFramework="net45" />
   <package id="Microsoft.Net.Http" version="2.2.28" targetFramework="net45" />
+  <package id="Microsoft.Rest.ClientRuntime" version="2.3.10" targetFramework="net452" />
```
Comment (maddieclayton) [inline src/ResourceManager/Maps/Maps.Test/packages.config]: Remove this.

File: src/ResourceManager/Maps/Maps.sln [hunk 4]
```diff
@@ -6,6 +6,8 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Commands.ResourceManager.Co
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Commands.Resources", "..\Resources\Commands.Resources\Commands.Resources.csproj", "{E1F5201D-6067-430E-B303-4E367652991B}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Commands.Tags", "..\Tags\Commands.Tags\Commands.Tags.csproj", "{2493A8F7-1949-4F29-8D53-9D459046C3B8}"
```
Comment (maddieclayton) [inline src/ResourceManager/Maps/Maps.sln]: Why was this added?

**Comment by maddieclayton** [comment]
@jp94 To answer questions from above:

No need to open a new PR in docs.  Once the release is live, the docs team will pull the docs from our repo. You just need to add the links (what you said about which links to add is correct).

The current PR removing the old module and adding the new simulataneous is much easier to review because it only shows the changes.  Please keep like this.

File: src/ResourceManager/Maps/Commands.Maps/help/AzureRM.Maps.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
+---
+Module Name: AzureRM.Maps
+Module Guid: bf60f35d-6c0b-42f2-be30-eb333a31279d
+Download Help Link: None
```
[the remark this answers is not in the document]
Comment (jp94) [reply src/ResourceManager/Maps/Commands.Maps/help/AzureRM.Maps.md]: Is it okay to reference the old locationbasedservices link over maps?

[the remark this answers is not in the document]
Comment (maddieclayton) [reply src/ResourceManager/Maps/Commands.Maps/help/AzureRM.Maps.md]: Oops no, please use https://docs.microsoft.com/en-us/powershell/module/azurerm.maps

File: src/ResourceManager/Maps/Maps.Test/TestController.cs (hunk 2 above)
Comment (jp94) [reply src/ResourceManager/Maps/Maps.Test/TestController.cs]: No changes were introduced.
If I recall, it worked ~2 months ago, but broke recently after pulling the latest code from the preview branch. This includes unmodified LocationBasedServices cmdlet tests, that yields the same error like shown below:
```
Test Name:	Microsoft.Azure.Commands.Maps.Test.ScenarioTests.MapsAccountTests.TestRemoveAccount
Test Outcome:	Failed
Result Message:	System.Management.Automation.ActionPreferenceStopException : The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: TestManagementClientHelper class wasn't initialized with the ResourceManagementClient client.
```

File: src/ResourceManager/Maps/Maps.sln (hunk 4 above)
Comment (jp94) [reply src/ResourceManager/Maps/Maps.sln]: Removed.

File: src/ResourceManager/Maps/AzureRM.Maps.psd1 [hunk 6]
```diff
@@ -63,20 +63,21 @@ RequiredAssemblies = '.\Microsoft.Azure.Management.LocationBasedServices.dll'
...
 # Cmdlets to export from this module, for best performance, do not use wildcards and do not delete the entry, use an empty array if there are no cmdlets to export.
-CmdletsToExport = 'Get-AzureRmLocationBasedServicesAccount', 
-               'New-AzureRmLocationBasedServicesAccount', 
-               'Remove-AzureRmLocationBasedServicesAccount', 
-               'Get-AzureRmLocationBasedServicesAccountKey', 
-               'New-AzureRmLocationBasedServicesAccountKey'
+CmdletsToExport = 'Get-AzureRmMapsAccount', 
+               'New-AzureRmMapsAccount', 
+               'Remove-AzureRmMapsAccount', 
+               'Set-AzureRmMapsAccount',
```
[the remark this answers is not in the document]
Comment (jp94) [reply src/ResourceManager/Maps/AzureRM.Maps.psd1]: That's interesting.
I see `Set-AzureRmLocationBasedServicesAccount` in [netcore](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/LocationBasedServices/AzureRM.LocationBasedServices.Netcore.psd1#L78), but not on the [original](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/LocationBasedServices/AzureRM.LocationBasedServices.psd1#L75) psd1. Should I remove that on both files?

File: src/ResourceManager/Maps/Maps.Test/ScenarioTests/MapsAccountTests.ps1 (hunk 1 above)
Comment (jp94) [reply src/ResourceManager/Maps/Maps.Test/ScenarioTests/MapsAccountTests.ps1]: Looks like the check is in under `function Test-RemoveAzureRmMapsAccount` at [line 73 & 83](https://github.com/jp94/azure-powershell/blob/c348a3e83ff522e26ed593d6cf9506c6fc219454/src/ResourceManager/Maps/Maps.Test/ScenarioTests/MapsAccountTests.ps1#L73). I think this check is good, but I can also add to all other test cases if needed.

File: src/ResourceManager/Maps/Maps.Test/ScenarioTests/MapsAccountTests.ps1 [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+{
+    # Setup
+    $rgname = Get-MapsManagementTestResourceName;
+
+    try
+    {
+        # Test
+        $accountname = 'ps-' + $rgname;
+        $skuname = 'S0';
+        $location = 'West US';
```
[the remark this answers is not in the document]
Comment (jp94) [reply src/ResourceManager/Maps/Maps.Test/ScenarioTests/MapsAccountTests.ps1]: Changing to `$location = Get-Location -providerNamespace 'Microsoft.Maps' -resourceType 'accounts' -preferredLocation 'West US';` or `Get-Location 'Microsoft.Maps' 'accounts' 'West US';` yields this error:
```
Test Name:	Microsoft.Azure.Commands.Maps.Test.ScenarioTests.MapsAccountTests.TestGetAccounts
Test Outcome:	Failed
Result Message:	System.Management.Automation.ActionPreferenceStopException : The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: Exception of type 'System.Management.Automation.PSInvalidOperationException' was thrown.
```

Perhaps I've mis-configured something? In `Maps.Test/bin/Debug/Common.ps1`, `Get-Location` function is there, but looks like it's failing to recognize it.

[the remark this answers is not in the document]
Comment (maddieclayton) [reply src/ResourceManager/Maps/Maps.Test/ScenarioTests/MapsAccountTests.ps1]: Hmm, I can't find anything in your code that would cause this error. My recommendation would be to do a git clean -xdf (after committing all local changes, as this will erase all local changes), rebuilding and rerunning the test.  I would think that something happened during all the renaming that is causing the function to not be found. Ping me if you are still having problems after this.

File: src/ResourceManager/Maps/Maps.Test/Properties/AssemblyInfo.cs [hunk 8]
```diff
@@ -47,6 +47,6 @@
 // You can specify all the values or you can default the Build and Revision Numbers 
 // by using the '*' as shown below:
 
-[assembly: AssemblyVersion("1.0.0")]
-[assembly: AssemblyFileVersion("1.0.0")]
+[assembly: AssemblyVersion("2.0.0")]
```
Comment (maddieclayton) [inline src/ResourceManager/Maps/Maps.Test/Properties/AssemblyInfo.cs]: This should be 1.0.0 as well.

File: src/ResourceManager/Maps/Maps.sln [hunk 9]
```diff
@@ -56,6 +56,10 @@ Global
 		{E1F5201D-6067-430E-B303-4E367652991B}.Debug|Any CPU.Build.0 = Debug|Any CPU
 		{E1F5201D-6067-430E-B303-4E367652991B}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{E1F5201D-6067-430E-B303-4E367652991B}.Release|Any CPU.Build.0 = Release|Any CPU
+		{2493A8F7-1949-4F29-8D53-9D459046C3B8}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
```
Comment (maddieclayton) [inline src/ResourceManager/Maps/Maps.sln]: Remove these four lines as well - this is from the tags.common project.

**Comment by maddieclayton** [comment]
@jp94 Looks like the build is failing because your "Remove" cmdlet does not implement passthru.  To add this, simply add a switchparameter "PassThru", and when the parameter is specified, print a boolean indicating the success or failure of the removal.  Let me know if you have any questions about this.

**Comment by jp94** [comment]
@maddieclayton 
Sorry about that (forgot to run `msbuild build.proj` after merging with `Azure/preview`).
I want to confirm the changes below before pushing.
```csharp
    [...]
+   [Cmdlet(VerbsCommon.Remove, MapsAccountNounStr, DefaultParameterSetName = NameParameterSet, SupportsShouldProcess = true), OutputType(typeof(bool))]
    public class RemoveAzureMapsAccount : MapsAccountBaseCmdlet
    {
        [...]

+       [Parameter(Mandatory = false)]
+       public SwitchParameter PassThru { get; set; }

        public override void ExecuteCmdlet()
        {
            base.ExecuteCmdlet();

            RunCmdLet(() =>
            {
                [...]

                if (!string.IsNullOrEmpty(rgName)
                    && !string.IsNullOrEmpty(name)
                    && ShouldProcess(name, string.Format(CultureInfo.CurrentCulture, Resources.RemoveAccount_ProcessMessage, name)))
                {
                    this.MapsClient.Accounts.Delete(rgName, name);
+                   if (PassThru.IsPresent)
+                   {
+                       WriteObject(true);
+                   }
                }
            });
        }
    }
}
```
I was able to complete the build without errors, and you can see the changes noted by `+` on the left side.
[Original](https://github.com/jp94/azure-powershell/blob/maps/src/ResourceManager/Maps/Commands.Maps/MapsAccount/RemoveAzureMapsAccount.cs)
Thanks.

**Comment by maddieclayton** [comment]
@jp94 You will also need to regenerate the help file, so the new parameter gets picked up. Thanks!

File: src/ResourceManager/Maps/Commands.Maps/help/Remove-AzureRmMapsAccount.md [hunk 10]
```diff
@@ -106,6 +106,19 @@ Accept pipeline input: True (ByPropertyName)
 Accept wildcard characters: False
 ```
 
+### -PassThru
+Return whether the specified account was successfully deleted or not.```yaml
```
[the remark this answers is not in the document]
Comment (jp94) [reply src/ResourceManager/Maps/Commands.Maps/help/Remove-AzureRmMapsAccount.md]: Yep.. I fixed the issue, and am running the whole `msbuild build.proj` again right now.

**Comment by maddieclayton** [comment]
on demand here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/871/
